### PR TITLE
Persist chroma db

### PR DESCRIPTION
**Pull request by ghost**
Description:
I have made changes to the Chroma DB Vector Store Wrapper to accept a persistent directory from client settings.

Issue: 
https://github.com/hwchase17/langchain/issues/6938

Dependencies:
None

Tag maintainer: @rlancemartin, @eyurtsev

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
seems to have combined spacy embeddings as well?

**Comment by rjarun8** [comment]
> seems to have combined spacy embeddings as well?

I guess so. my bad. let me correct that.

**Comment by hwchase17** [comment]
still has some conflicts...

**Comment by rjarun8** [comment]
> > > 
> > 
> > 
> > The conflicts are resolved. Getting linting issues. However, in my space, the checks are passing.  Sorry, am I missing something?
> > ```
> > 
> > 
> > 
> > !poetry run black chroma.py spacy_embedding.py --check
> > 
> > 
> > 
> > 
> > 
> > All done! ✨ 🍰 ✨
> > 
> > 2 files would be left unchanged
> > ```
> 
> Please run "make format"

Thanks . resolved. but there is another blocker. An integration test is failing. But I can see that in **_https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/chroma.py_** the _**similarity_search_by_vector_with_relevance_scores**_ is indeed available. what else could be causing it?

```
poetry run mypy .
tests/integration_tests/vectorstores/test_chroma.py:74: error: "Chroma" has no attribute "similarity_search_by_vector_with_relevance_scores"; maybe "_similarity_search_with_relevance_scores"?  [attr-defined]
Found 1 error in 1 file (checked 1288 source files)
make: *** [Makefile:34: lint] Error 1
Error: Process completed with exit code 2.
```

I also tested it locally , same error.

```
"""Test Chroma functionality."""
import pytest

from langchain.docstore.document import Document
from langchain.vectorstores import Chroma
from langchain.embeddings import FakeEmbeddings

def test_chroma_with_metadatas_with_scores_using_vector() -> None:
    """Test end to end construction and scored search, using embedding vector."""
    texts = ["foo", "bar", "baz"]
    metadatas = [{"page": str(i)} for i in range(len(texts))]
    embeddings = FakeEmbeddings(size=5)

    docsearch = Chroma.from_texts(
        collection_name="test_collection",
        texts=texts,
        embedding=embeddings,
        metadatas=metadatas,
    )
    embedded_query = embeddings.embed_query("foo")
    output = docsearch.similarity_search_by_vector_with_relevance_scores(
        embedding=embedded_query, k=1
    )
    assert output == [(Document(page_content="foo", metadata={"page": "0"}), 0.0)]

test_chroma_with_metadatas_with_scores_using_vector()

```




```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-12-7062ecfb5130>](https://localhost:8080/#) in <cell line: 1>()
----> 1 test_chroma_with_metadatas_with_scores_using_vector()

[<ipython-input-11-87e2c68a20e2>](https://localhost:8080/#) in test_chroma_with_metadatas_with_scores_using_vector()
     12     )
     13     embedded_query = embeddings.embed_query("foo")
---> 14     output = docsearch.similarity_search_by_vector_with_relevance_scores(
     15         embedding=embedded_query, k=1
     16     )

AttributeError: 'Chroma' object has no attribute 'similarity_search_by_vector_with_relevance_scores'
```

**Comment by rlancemartin** [comment]
> > > > 
> 
> > > 
> 
> > > 
> 
> > > The conflicts are resolved. Getting linting issues. However, in my space, the checks are passing.  Sorry, am I missing something?
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > > 
> 
> > > !poetry run black chroma.py spacy_embedding.py --check
> 
> > > 
> 
> > > 
> 
> > > 
> 
> > > 
> 
> > > 
> 
> > > All done! ✨ 🍰 ✨
> 
> > > 
> 
> > > 2 files would be left unchanged
> 
> > > ```
> 
> > 
> 
> > Please run "make format"
> 
> 
> 
> Thanks . resolved. but there is another blocker. An integration test is failing. But I can see that in **_https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/chroma.py_** the _**similarity_search_by_vector_with_relevance_scores**_ is indeed available. what else could be causing it?
> 
> 
> 
> ```
> 
> poetry run mypy .
> 
> tests/integration_tests/vectorstores/test_chroma.py:74: error: "Chroma" has no attribute "similarity_search_by_vector_with_relevance_scores"; maybe "_similarity_search_with_relevance_scores"?  [attr-defined]
> 
> Found 1 error in 1 file (checked 1288 source files)
> 
> make: *** [Makefile:34: lint] Error 1
> 
> Error: Process completed with exit code 2.
> 
> ```
> 
> 
> 
> I also tested it locally , same error.
> 
> 
> 
> ```
> 
> """Test Chroma functionality."""
> 
> import pytest
> 
> 
> 
> from langchain.docstore.document import Document
> 
> from langchain.vectorstores import Chroma
> 
> from langchain.embeddings import FakeEmbeddings
> 
> 
> 
> def test_chroma_with_metadatas_with_scores_using_vector() -> None:
> 
>     """Test end to end construction and scored search, using embedding vector."""
> 
>     texts = ["foo", "bar", "baz"]
> 
>     metadatas = [{"page": str(i)} for i in range(len(texts))]
> 
>     embeddings = FakeEmbeddings(size=5)
> 
> 
> 
>     docsearch = Chroma.from_texts(
> 
>         collection_name="test_collection",
> 
>         texts=texts,
> 
>         embedding=embeddings,
> 
>         metadatas=metadatas,
> 
>     )
> 
>     embedded_query = embeddings.embed_query("foo")
> 
>     output = docsearch.similarity_search_by_vector_with_relevance_scores(
> 
>         embedding=embedded_query, k=1
> 
>     )
> 
>     assert output == [(Document(page_content="foo", metadata={"page": "0"}), 0.0)]
> 
> 
> 
> test_chroma_with_metadatas_with_scores_using_vector()
> 
> 
> 
> ```
> 
> 
> 
> 
> 
> 
> 
> 
> 
> ```
> 
> ---------------------------------------------------------------------------
> 
> AttributeError                            Traceback (most recent call last)
> 
> [<ipython-input-12-7062ecfb5130>](https://localhost:8080/#) in <cell line: 1>()
> 
> ----> 1 test_chroma_with_metadatas_with_scores_using_vector()
> 
> 
> 
> [<ipython-input-11-87e2c68a20e2>](https://localhost:8080/#) in test_chroma_with_metadatas_with_scores_using_vector()
> 
>      12     )
> 
>      13     embedded_query = embeddings.embed_query("foo")
> 
> ---> 14     output = docsearch.similarity_search_by_vector_with_relevance_scores(
> 
>      15         embedding=embedded_query, k=1
> 
>      16     )
> 
> 
> 
> AttributeError: 'Chroma' object has no attribute 'similarity_search_by_vector_with_relevance_scores'
> 
> ```

Did you re-build Langchain?

**Comment by rjarun8** [comment]
> 

Can you explain that? I did quite catch that. I used the source code via pip and ran the test module locally to see if it was able to fetch. But weirdly it doesn't seem to.

**Comment by rlancemartin** [comment]
> > 
> 
> 
> 
> Can you explain that? I did quite catch it. I used the source code via pip and ran the test module locally to see if it was able to fetch. But weirdly it doesn't seem to.

I'll have a close look today.

File: langchain/vectorstores/chroma.py [hunk 1]
```diff
@@ -192,31 +198,6 @@ def similarity_search_by_vector(
         )
         return _results_to_docs(results)
 
-    def similarity_search_by_vector_with_relevance_scores(
```
Comment (rlancemartin) [inline langchain/vectorstores/chroma.py:195]: why did you remove this method? this is why test is failing.

Comment (rjarun8) [reply langchain/vectorstores/chroma.py]: > why did you remove this method? this is why test is failing.

@rlancemartin I am not sure. Totally unintended. Thanks for pointing it out.

**Comment by rlancemartin** [comment]
@rjarun8 this PR has gotten into a slightly confusing state w/ a large number of commits and apparently unrelated changes that have been pulled in. do you want to simply put up a fresh PR on current `master` with the precise changes you intent to merge?

**Review by rlancemartin — CHANGES_REQUESTED** [review]
Please put up a fresh PR w/ the precise changes you want to merge.

**Comment by rjarun8** [comment]
> @rjarun8 this PR has gotten into a slightly confusing state w/ a large number of commits and apparently unrelated changes that have been pulled in. do you want to simply put up a fresh PR on current `master` with the precise changes you intent to merge?

Sure @rlancemartin that would be better I guess to evade this confusion created. sorry about that.

**Comment by rlancemartin** [comment]
> > @rjarun8 this PR has gotten into a slightly confusing state w/ a large number of commits and apparently unrelated changes that have been pulled in. do you want to simply put up a fresh PR on current `master` with the precise changes you intent to merge?
> 
> Sure @rlancemartin that would be better I guess to evade this confusion created. sorry about that.

Thanks! Np. Sometimes easier just to start fresh :)

**Pull request closed without merging**